src/scenic/core/propositions.py

```diff
@@ -0,0 +1,193 @@
+"""Objects representing propositions that can be used to specify conditions"""
+
+from functools import reduce
+import operator
+from typing import List
+
+import rv_ltl
+
+from scenic.core.errors import InvalidScenarioError
+from scenic.core.lazy_eval import needsLazyEvaluation
+
+class PropositionMonitor:
+    def __init__(self, proposition: "PropositionNode") -> None:
+        self._proposition = proposition
+        self._monitor = proposition.ltl_node.create_monitor()
+        
+    def update(self):
+        atomic_propositions = self._proposition.atomics()
+        state = {}
+        for ap in atomic_propositions:
+            b = ap.closure()
+            if needsLazyEvaluation(b):
+                raise InvalidScenarioError(f'value undefined outside of object definition')
+            state[str(ap.syntax_id)] = b
+        self._monitor.update(state)
+        return self._monitor.evaluate()
+
+
+class PropositionNode:
+    """Base class for temporal and non-temporal propositions"""
+    def __init__(self, ltl_node) -> None:
+        self.ltl_node = ltl_node
+
+        self.is_temporal = False
+        """tells if the proposition is temporal"""
+
+    def check_constrains_sampling(self):
+        """checks if the proposition can be used for pruning
+        A requirement can be used for pruning if it is evaluated on the scene generation phase before simulation, and
+        violation in that phase immediately results in discarding the scene and regenerating a new one.
+        For simplicity, we currently check two special cases:
+        1. requirements with no temporal requirement
+        2. requirements with only one `always` operator on top-level
+        Returns:
+            bool: True if the requirement is one of the forms above. False otherwise.
+        """
```

Comment:
@shumbo This docstring doesn't have the right form (if you look at the compiled documentation you'll see it doesn't show up correctly). There needs to be a blank line between the 1-line summary and the rest, and I think the `Returns:` section also needs a blank line before it. I won't comment on the other docstrings in your code but it would be good to make a pass through all of them to make sure they appear correctly in the docs.

Reply:
@shumbo I fixed the docstrings in this file, but if you want to look at the other docstrings you wrote that would be great.

Reply:
Thanks, Eric!